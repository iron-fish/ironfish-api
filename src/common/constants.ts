/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { EventType } from '.prisma/client';

export const DAYS_IN_WEEK = 7;
export const MS_PER_DAY = 1000 * 60 * 60 * 24;

// Pagination limits
export const DEFAULT_LIMIT = 20;
export const MAX_LIMIT = 100;

// Event categories
export const WEEKLY_POINT_LIMITS_BY_EVENT_TYPE: Record<EventType, number> = {
  [EventType.BLOCK_MINED]: 1000,
  [EventType.BUG_CAUGHT]: 1000,
  [EventType.COMMUNITY_CONTRIBUTION]: 2000,
  [EventType.PULL_REQUEST_MERGED]: 5000,
  [EventType.SOCIAL_MEDIA_PROMOTION]: 1000,
};

export const POINTS_PER_CATEGORY: Record<EventType, number> = {
  [EventType.BLOCK_MINED]: 100,
  [EventType.BUG_CAUGHT]: 100,
  [EventType.COMMUNITY_CONTRIBUTION]: 1000,
  [EventType.PULL_REQUEST_MERGED]: 500,
  [EventType.SOCIAL_MEDIA_PROMOTION]: 100,
};
