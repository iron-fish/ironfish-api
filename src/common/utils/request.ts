/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Request } from 'express';

export function fetchIpAddressFromRequest(request: Request): string {
  const xForwardedFor = request.header('X-Forwarded-For');
  if (xForwardedFor) {
    const addresses = xForwardedFor.split(',');
    if (addresses.length) {
      return addresses[0].trim();
    }
  }

  // In Heroku, `X-Forwarded-For` will always be set. However, this will be used
  // as a fallback in local development.
  return request.ip;
}
