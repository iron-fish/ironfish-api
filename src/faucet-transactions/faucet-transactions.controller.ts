/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import {
  Body,
  Controller,
  Get,
  HttpCode,
  HttpStatus,
  NotFoundException,
  Param,
  ParseIntPipe,
  Post,
  UseGuards,
  ValidationPipe,
} from '@nestjs/common';
import {
  ApiExcludeEndpoint,
  ApiOperation,
  ApiParam,
  ApiTags,
} from '@nestjs/swagger';
import { ApiKeyGuard } from '../auth/guards/api-key.guard';
import { CompleteFaucetTransactionDto } from './dto/complete-faucet-transaction.dto';
import { CreateFaucetTransactionDto } from './dto/create-faucet-transaction.dto';
import { FaucetTransactionsService } from './faucet-transactions.service';
import { FaucetTransactionsStatus } from './interfaces/faucet-transactions-status';
import { SerializedFaucetTransaction } from './interfaces/serialized-faucet-transaction';
import { serializedFaucetTransactionFromRecord } from './utils/faucet-transactions.translator';

@ApiTags('Faucet Transactions')
@Controller('faucet_transactions')
export class FaucetTransactionsController {
  constructor(
    private readonly faucetTransactionsService: FaucetTransactionsService,
  ) {}

  @ApiExcludeEndpoint()
  @Post()
  async create(
    @Body(
      new ValidationPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
      }),
    )
    { email, public_key: publicKey }: CreateFaucetTransactionDto,
  ): Promise<SerializedFaucetTransaction> {
    return serializedFaucetTransactionFromRecord(
      await this.faucetTransactionsService.create({ email, publicKey }),
    );
  }

  @ApiExcludeEndpoint()
  @Get('next')
  @UseGuards(ApiKeyGuard)
  async next(): Promise<SerializedFaucetTransaction> {
    const nextFaucetTransaction = await this.faucetTransactionsService.next();
    if (!nextFaucetTransaction) {
      throw new NotFoundException();
    }
    return serializedFaucetTransactionFromRecord(nextFaucetTransaction);
  }

  @ApiOperation({ summary: 'Returns the global status of faucet transactions' })
  @Get('status')
  async status(): Promise<FaucetTransactionsStatus> {
    return this.faucetTransactionsService.getGlobalStatus();
  }

  @ApiOperation({ summary: 'Gets a specific Faucet Transaction' })
  @ApiParam({ description: 'Unique Faucet Transaction identifier', name: 'id' })
  @Get(':id')
  async find(
    @Param(
      'id',
      new ParseIntPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
      }),
    )
    id: number,
  ): Promise<SerializedFaucetTransaction> {
    const record = await this.faucetTransactionsService.findOrThrow(id);
    return serializedFaucetTransactionFromRecord(record);
  }

  @ApiExcludeEndpoint()
  @Post(':id/start')
  @HttpCode(HttpStatus.OK)
  @UseGuards(ApiKeyGuard)
  async start(
    @Param(
      'id',
      new ParseIntPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
      }),
    )
    id: number,
  ): Promise<SerializedFaucetTransaction> {
    const record = await this.faucetTransactionsService.findOrThrow(id);
    return serializedFaucetTransactionFromRecord(
      await this.faucetTransactionsService.start(record),
    );
  }

  @ApiExcludeEndpoint()
  @Post(':id/complete')
  @HttpCode(HttpStatus.OK)
  @UseGuards(ApiKeyGuard)
  async complete(
    @Param(
      'id',
      new ParseIntPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
      }),
    )
    id: number,
    @Body(
      new ValidationPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
      }),
    )
    dto: CompleteFaucetTransactionDto,
  ): Promise<SerializedFaucetTransaction> {
    const record = await this.faucetTransactionsService.findOrThrow(id);
    return serializedFaucetTransactionFromRecord(
      await this.faucetTransactionsService.complete(record, dto),
    );
  }
}
