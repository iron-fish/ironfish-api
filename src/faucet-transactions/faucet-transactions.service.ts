/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import {
  Injectable,
  NotFoundException,
  UnprocessableEntityException,
} from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { CompleteFaucetTransactionOptions } from './interfaces/complete-faucet-transaction-options';
import { CreateFaucetTransactionOptions } from './interfaces/create-faucet-transaction-options';
import { FaucetTransactionsStatus } from './interfaces/faucet-transactions-status';
import { FaucetTransaction, Prisma } from '.prisma/client';

export const FAUCET_REQUESTS_LIMIT = 3;

@Injectable()
export class FaucetTransactionsService {
  constructor(private readonly prisma: PrismaService) {}

  async findOrThrow(id: number): Promise<FaucetTransaction> {
    const record = await this.prisma.faucetTransaction.findUnique({
      where: {
        id,
      },
    });
    if (!record) {
      throw new NotFoundException();
    }
    return record;
  }

  async create({
    email,
    publicKey,
  }: CreateFaucetTransactionOptions): Promise<FaucetTransaction> {
    return this.prisma.$transaction(async (prisma) => {
      const count = await this.prisma.faucetTransaction.count({
        where: {
          OR: [
            { email },
            {
              public_key: publicKey,
            },
          ],
        },
      });
      if (count >= FAUCET_REQUESTS_LIMIT) {
        throw new UnprocessableEntityException({
          code: 'faucet_max_requests_reached',
          message: 'Too many faucet requests',
        });
      }
      return prisma.faucetTransaction.create({
        data: {
          email,
          public_key: publicKey,
        },
      });
    });
  }

  async next(): Promise<FaucetTransaction | null> {
    return this.prisma.$transaction(async (prisma) => {
      const currentlyRunningFaucetTransaction =
        await prisma.faucetTransaction.findFirst({
          where: {
            started_at: {
              not: null,
            },
            completed_at: null,
          },
          orderBy: {
            created_at: Prisma.SortOrder.asc,
          },
        });
      if (currentlyRunningFaucetTransaction) {
        return currentlyRunningFaucetTransaction;
      }
      return prisma.faucetTransaction.findFirst({
        where: {
          started_at: null,
          completed_at: null,
        },
        orderBy: {
          created_at: Prisma.SortOrder.asc,
        },
      });
    });
  }

  async start(
    faucetTransaction: FaucetTransaction,
  ): Promise<FaucetTransaction> {
    if (faucetTransaction.completed_at) {
      throw new UnprocessableEntityException();
    }
    return this.prisma.$transaction(async (prisma) => {
      return prisma.faucetTransaction.update({
        data: {
          started_at: new Date().toISOString(),
          tries: {
            increment: 1,
          },
        },
        where: {
          id: faucetTransaction.id,
        },
      });
    });
  }

  async complete(
    faucetTransaction: FaucetTransaction,
    options?: CompleteFaucetTransactionOptions,
  ): Promise<FaucetTransaction> {
    if (faucetTransaction.completed_at) {
      throw new UnprocessableEntityException();
    }
    return this.prisma.$transaction(async (prisma) => {
      return prisma.faucetTransaction.update({
        data: {
          completed_at: new Date().toISOString(),
          hash: options?.hash,
        },
        where: {
          id: faucetTransaction.id,
        },
      });
    });
  }

  async getGlobalStatus(): Promise<FaucetTransactionsStatus> {
    const [pending, running, completed] = await this.prisma.$transaction([
      this.prisma.faucetTransaction.count({
        where: {
          started_at: null,
          completed_at: null,
        },
      }),
      this.prisma.faucetTransaction.count({
        where: {
          started_at: {
            not: null,
          },
          completed_at: null,
        },
      }),
      this.prisma.faucetTransaction.count({
        where: {
          completed_at: {
            not: null,
          },
        },
      }),
    ]);
    return {
      completed,
      running,
      pending,
    };
  }
}
