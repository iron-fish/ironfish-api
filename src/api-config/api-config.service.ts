/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';

@Injectable()
export class ApiConfigService {
  constructor(private readonly config: ConfigService) {}

  get<T = unknown>(name: string): T {
    const value = this.config.get<T>(name);
    if (value === undefined) {
      throw new Error(`Missing config value for ${name}`);
    }
    return value;
  }

  getWithDefault<T = unknown>(name: string, defaultValue: T): T {
    const value = this.config.get<T>(name);
    if (value === undefined) {
      return defaultValue;
    }
    return value;
  }

  isLocal(): boolean {
    return this.get<string>('NODE_ENV') === 'development';
  }

  isStaging(): boolean {
    return this.get<string>('NODE_ENV') === 'staging';
  }

  isProduction(): boolean {
    return this.get<string>('NODE_ENV') === 'production';
  }

  get dbPoolUrl(): string {
    return `${this.get<string>('DATABASE_CONNECTION_POOL_URL')}?pgbouncer=true`;
  }
}
