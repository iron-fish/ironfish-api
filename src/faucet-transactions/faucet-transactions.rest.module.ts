/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Module } from '@nestjs/common';
import { ApiConfigModule } from '../api-config/api-config.module';
import { FaucetTransactionsController } from './faucet-transactions.controller';
import { FaucetTransactionsModule } from './faucet-transactions.module';

@Module({
  controllers: [FaucetTransactionsController],
  imports: [ApiConfigModule, FaucetTransactionsModule],
})
export class FaucetTransactionsRestModule {}
