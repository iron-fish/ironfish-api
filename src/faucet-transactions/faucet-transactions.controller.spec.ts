/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { HttpStatus, INestApplication } from '@nestjs/common';
import faker from 'faker';
import request from 'supertest';
import { ulid } from 'ulid';
import { ApiConfigService } from '../api-config/api-config.service';
import { bootstrapTestApp } from '../test/test-app';
import { FaucetTransactionsService } from './faucet-transactions.service';

describe('FaucetTransactionsController', () => {
  let app: INestApplication;
  let faucetTransactionsService: FaucetTransactionsService;

  let API_KEY: string;

  beforeAll(async () => {
    app = await bootstrapTestApp();
    faucetTransactionsService = app.get(FaucetTransactionsService);
    API_KEY = app.get(ApiConfigService).get<string>('IRONFISH_API_KEY');
    await app.init();
  });

  afterAll(async () => {
    await app.close();
  });

  describe('POST /faucet_transactions', () => {
    describe('with missing arguments', () => {
      it('returns a 422', async () => {
        const { body } = await request(app.getHttpServer())
          .post('/faucet_transactions')
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with empty arguments', () => {
      it('returns a 422', async () => {
        const { body } = await request(app.getHttpServer())
          .post('/faucet_transactions')
          .send({ email: '', public_key: '' })
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with valid arguments', () => {
      it('creates a Faucet Transaction', async () => {
        const publicKey = ulid();
        const { body } = await request(app.getHttpServer())
          .post('/faucet_transactions')
          .send({ email: faker.internet.email(), public_key: publicKey })
          .expect(HttpStatus.CREATED);

        expect(body).toMatchObject({
          object: 'faucet_transaction',
          id: expect.any(Number),
          public_key: publicKey,
        });
      });
    });
  });

  describe('GET /faucet_transactions/next', () => {
    describe('with a missing api key', () => {
      it('returns a 401', async () => {
        const { body } = await request(app.getHttpServer())
          .get('/faucet_transactions/next')
          .expect(HttpStatus.UNAUTHORIZED);

        expect(body).toMatchSnapshot();
      });
    });

    describe('when no next FaucetTransaction is available', () => {
      it('returns a 404', async () => {
        jest
          .spyOn(faucetTransactionsService, 'next')
          .mockResolvedValueOnce(null);

        await request(app.getHttpServer())
          .get('/faucet_transactions/next')
          .set('Authorization', `Bearer ${API_KEY}`)
          .expect(HttpStatus.NOT_FOUND);
      });
    });

    describe('when there is a pending FaucetTransaction', () => {
      it('returns the record', async () => {
        jest.spyOn(faucetTransactionsService, 'next').mockResolvedValueOnce({
          id: 0,
          created_at: new Date(),
          updated_at: new Date(),
          public_key: 'mock-key',
          email: null,
          completed_at: null,
          started_at: null,
        });

        const { body } = await request(app.getHttpServer())
          .get('/faucet_transactions/next')
          .set('Authorization', `Bearer ${API_KEY}`)
          .expect(HttpStatus.OK);

        expect(body).toMatchObject({
          object: 'faucet_transaction',
          id: expect.any(Number),
          public_key: expect.any(String),
        });
      });
    });
  });

  describe('POST /faucet_transactions/:id/start', () => {
    describe('with a missing api key', () => {
      it('returns a 401', async () => {
        const { body } = await request(app.getHttpServer())
          .post('/faucet_transactions/0/start')
          .expect(HttpStatus.UNAUTHORIZED);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with an invalid id', () => {
      it('returns a 404', async () => {
        await request(app.getHttpServer())
          .get('/faucet_transactions/100000/start')
          .set('Authorization', `Bearer ${API_KEY}`)
          .expect(HttpStatus.NOT_FOUND);
      });
    });

    describe('with a valid id', () => {
      it('starts the FaucetTransaction', async () => {
        const email = faker.internet.email();
        const publicKey = ulid();
        const faucetTransaction = await faucetTransactionsService.create({
          email,
          publicKey,
        });

        const { body } = await request(app.getHttpServer())
          .post(`/faucet_transactions/${faucetTransaction.id}/start`)
          .set('Authorization', `Bearer ${API_KEY}`)
          .expect(HttpStatus.OK);

        expect(body).toMatchObject({
          object: 'faucet_transaction',
          id: faucetTransaction.id,
          started_at: expect.any(String),
        });
      });
    });
  });

  describe('POST /faucet_transactions/:id/complete', () => {
    describe('with a missing api key', () => {
      it('returns a 401', async () => {
        const { body } = await request(app.getHttpServer())
          .post('/faucet_transactions/0/complete')
          .expect(HttpStatus.UNAUTHORIZED);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with an invalid id', () => {
      it('returns a 404', async () => {
        await request(app.getHttpServer())
          .get('/faucet_transactions/100000/complete')
          .set('Authorization', `Bearer ${API_KEY}`)
          .expect(HttpStatus.NOT_FOUND);
      });
    });

    describe('with a valid id', () => {
      it('starts the FaucetTransaction', async () => {
        const email = faker.internet.email();
        const publicKey = ulid();
        const faucetTransaction = await faucetTransactionsService.create({
          email,
          publicKey,
        });
        await faucetTransactionsService.start(faucetTransaction);

        const { body } = await request(app.getHttpServer())
          .post(`/faucet_transactions/${faucetTransaction.id}/complete`)
          .set('Authorization', `Bearer ${API_KEY}`)
          .expect(HttpStatus.OK);

        expect(body).toMatchObject({
          object: 'faucet_transaction',
          id: faucetTransaction.id,
          completed_at: expect.any(String),
          started_at: expect.any(String),
        });
      });
    });
  });
});
