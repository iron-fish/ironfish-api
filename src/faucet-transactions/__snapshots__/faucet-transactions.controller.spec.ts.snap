// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FaucetTransactionsController GET /faucet_transactions/next with a missing api key returns a 401 1`] = `
Object {
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`FaucetTransactionsController POST /faucet_transactions with empty arguments returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": Array [
    "email must be an email",
    "public_key should not be empty",
  ],
  "statusCode": 422,
}
`;

exports[`FaucetTransactionsController POST /faucet_transactions with missing arguments returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": Array [
    "public_key must be a string",
    "public_key should not be empty",
  ],
  "statusCode": 422,
}
`;

exports[`FaucetTransactionsController POST /faucet_transactions/:id/complete with a missing api key returns a 401 1`] = `
Object {
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`FaucetTransactionsController POST /faucet_transactions/:id/start with a missing api key returns a 401 1`] = `
Object {
  "message": "Unauthorized",
  "statusCode": 401,
}
`;
